### PR TITLE
Hashicorp Vault Integration

### PR DESCRIPTION
**Pull request by makemp**

### CODE REVIEW

File: lib/travis/build/appliances/vault_connect.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      class VaultConnect < Base
+        ERROR_MESSAGE = ["Failed to connect to the Vault instance. Please verify if:\n* The Vault Token is correct. \n* The Vault token is not expired. \n* The Vault can accept connections from the Travis CI build job environments (https://docs.travis-ci.com/user/ip-addresses/).", ansi: :red]
+        SUCCESS_MESSAGE = ['Connected to Vault instance.', ansi: :green]
+
+        def apply?
+          @vault = config[:vault]
+        end
+
+        def apply
+          ENV['VAULT_ADDR'] = @vault[:api_url]
```
Comment (makemp) [inline lib/travis/build/appliances/vault_connect.rb:16]: Is that you had on mind when you told me abbout setting VAULT_ADDR and VAULT_TOKEN variables or I just invoke `sh.export` here?

File: spec/fixtures/build_config_with_vault_kv2.json [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    "number": "18939"
+  },
+  "ssh_key": null,
+  "timeouts": {
+    "hard_limit": null,
+    "log_silence": null
+  },
+  "type": "test",
+  "vm_type": "default"
+}
```
Comment (vitalied) [inline spec/fixtures/build_config_with_vault_kv2.json:114]: new line at the end of file

**Review by vitalied — CHANGES_REQUESTED** [review]

File: lib/travis/build/appliances/vault_connect.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+          ENV['VAULT_TOKEN'] = @vault[:token]
+          Travis::Vault::Connect.call
+          sh.echo *SUCCESS_MESSAGE
+        rescue Travis::Vault::ConnectionError, ArgumentError, URI::InvalidURIError => _e
+          sh.echo *ERROR_MESSAGE
+          sh.terminate
```
[the remark this answers is not in the document]
Comment (makemp) [reply lib/travis/build/appliances/vault_connect.rb:22]: That was I am asking:
https://github.com/travis-ci/travis-build/pull/2033#discussion_r914669848
Should I use export in shell or do this with ENV?
Once the builder did its job the instance is purged, right? so All ENV's are gone.
I just do not know the purposes of this VAULT_ADDR and VAULT_TOKEN, what do you plan to do with these afterwards so I do not know if I should leave it as it is now or export it into shell

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -5,6 +5,9 @@
 require 'sinatra/test_helpers'
 require 'travis/build'
 require 'pathname'
+require 'webmock/rspec'
+
+WebMock.allow_net_connect!
```
Comment (vitalied) [inline spec/spec_helper.rb:10]: @makemp why to allow net connections?

File: lib/travis/services/vault/keys/resolver.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+          end
+        end
+
+        private
+
+        attr_reader :paths, :version, :appliance
+
+        delegate :export, :echo, to: 'appliance.sh'
+
+        def warn_message(path)
```
[the remark this answers is not in the document]
Comment (makemp) [reply lib/travis/services/vault/keys/resolver.rb:30]: I can't move this to constant, as it has an argument (path)

File: lib/travis/services/vault/keys/kv2.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  module Vault
+    class Keys
+      class KV2
+        def self.resolve(path, vault)
+          faraday_connection = Faraday.new(
+            url: vault[:api_url],
+            headers: { 'X-Vault-Token' => vault[:token] }
+          )
+          response = faraday_connection.get("/v1/secret/data/#{path}")
+          response.status == 200 ? JSON.parse(response.body).dig('data', 'data').to_json : nil
```
Comment (vitalied) [inline lib/travis/services/vault/keys/kv2.rb:11]: `JSON.parse(response.body).dig('data', 'data').to_json if response.status == 200`

File: lib/travis/services/vault/connect.rb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+module Travis
+  module Vault
+    class Connect
+      def self.call(vault)
+        faraday_connection = Faraday.new(
+          url: vault[:api_url],
+          headers: { 'X-Vault-Token' => vault[:token] }
```
Comment (vitalied) [inline lib/travis/services/vault/connect.rb:7]: `headers: { 'X-Vault-Token': vault[:token] }`

File: lib/travis/services/vault/keys/kv1.rb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+module Travis
+  module Vault
+    class Keys
+      class KV1
+        def self.resolve(path, vault)
+          faraday_connection = Faraday.new(
+            url: vault[:api_url],
+            headers: { 'X-Vault-Token' => vault[:token] }
```
Comment (vitalied) [inline lib/travis/services/vault/keys/kv1.rb:8]: `headers: { 'X-Vault-Token': vault[:token] }`

File: lib/travis/services/vault/keys/kv2.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+module Travis
+  module Vault
+    class Keys
+      class KV2
+        def self.resolve(path, vault)
+          faraday_connection = Faraday.new(
+            url: vault[:api_url],
+            headers: { 'X-Vault-Token' => vault[:token] }
```
Comment (vitalied) [inline lib/travis/services/vault/keys/kv2.rb:8]: `headers: { 'X-Vault-Token': vault[:token] }`

File: spec/build/services/vault/connect_spec.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+      {
+        api_url: 'https://myvault.org',
+        token: 'my-token'
+      }
+    end
+
+    context 'the endpoint returns 200' do
+      before do
+        stub_request(:get, 'https://myvault.org/v1/auth/token/lookup-self').
+          with(headers: { 'X-Vault-Token' => 'my-token' }).
```
Comment (vitalied) [inline spec/build/services/vault/connect_spec.rb:17]: `headers: { 'X-Vault-Token': 'my-token' }`
replace everywhere

**Review by vitalied — CHANGES_REQUESTED** [review]

**Pull request merged**